### ISSUE TITLE
assert failure

### ISSUE THREAD

**Issue by dtaht** [jech/babeld#43]
I got around to briefly testing jech head today. It blew up with an assert failure. I did not look into it harder, but can if needed.

...

````
My id ee:a8:6b:ff:fe:fe:09:a2 seqno 38769
Neighbour fe80::f6f2:6dff:feb6:a01c dev eno1 reach e000 ureach 0000 rxcost 96 txcost 96 rtt 3.868 rttcost 0 chan -2.
Neighbour fe80::46d9:e7ff:fe93:822e dev eno1 reach e000 ureach 0000 rxcost 96 txcost 96 rtt 1.634 rttcost 0 chan -2.
Neighbour fe80::e091:f5ff:febe:a353 dev eno1 reach e000 ureach 0000 rxcost 96 txcost 96 rtt 1.154 rttcost 0 chan -2.
Neighbour fe80::225:90ff:fec1:6252 dev eno1 reach e000 ureach 0000 rxcost 96 txcost 96 rtt 1.010 rttcost 0 chan -2.
Neighbour fe80::225:90ff:fec2:2aa3 dev eno1 reach e000 ureach 0000 rxcost 96 txcost 96 rtt 11.178 rttcost 0 chan -2.
::/0 metric 0 (exported)
2603:3024:1536:86f0::/64 metric 0 (exported)
172.22.0.0/24 from 0.0.0.0/0 metric 0 (exported)
::/0 from 2603:3024:1536:8600:f6f2:6dff:feb6:a01d/128 metric 192 (65535) refmetric 96 id f6:f2:6d:ff:fe:b6:a0:1d seqno 45242 age 1 via eno1 neigh fe80::225:90ff:fec1:6252 (installed)
::/0 from 2603:3024:1536:86f0::/60 metric 192 (65535) refmetric 96 id f6:f2:6d:ff:fe:b6:a0:1d seqno 45242 age 1 via eno1 neigh fe80::225:90ff:fec1:6252 (installed)
::/0 metric 2082 (65535) refmetric 1986 id f6:f2:6d:ff:fe:b6:a0:1d seqno 45242 age 1 via eno1 neigh fe80::225:90ff:fec2:2aa3 (feasible)
::/0 metric 2082 (65535) refmetric 1986 id f6:f2:6d:ff:fe:b6:a0:1d seqno 45242 age 1 via eno1 neigh fe80::225:90ff:fec1:6252 (feasible)
::/0 metric 2082 (65535) refmetric 1986 id f6:f2:6d:ff:fe:b6:a0:1d seqno 45242 age 1 via eno1 neigh fe80::46d9:e7ff:fe93:822e (feasible)
0.0.0.0/0 metric 192 (65535) refmetric 96 id f6:f2:6d:ff:fe:b6:a0:1d seqno 45242 age 1 via eno1 neigh fe80::225:90ff:fec1:6252 nexthop 172.22.0.172 (installed)
50.197.142.144/29 metric 809 (65535) refmetric 713 id 02:0d:b9:ff:fe:43:a0:6c seqno 51873 age 1 via eno1 neigh fe80::225:90ff:fec1:6252 nexthop 172.22.0.172 (installed)
172.20.0.0/14 metric 192 (65535) refmetric 96 id e2:91:f5:ff:fe:be:a3:54 seqno 38347 age 1 via eno1 neigh fe80::225:90ff:fec1:6252 nexthop 172.22.0.172 (installed)
172.22.220.0/22 metric 192 (65535) refmetric 96 id 46:d9:e7:ff:fe:93:82:2d seqno 27834 age 1 via eno1 neigh fe80::225:90ff:fec1:6252 nexthop 172.22.0.172 (installed)
2601:646:8301:6760::/60 metric 384 (65535) refmetric 288 id a2:21:b7:ff:fe:ac:e4:55 seqno 10074 age 1 via eno1 neigh fe80::225:90ff:fec1:6252 (installed)
2603:3024:1536:86f0::/60 metric 192 (65535) refmetric 96 id f6:f2:6d:ff:fe:b6:a0:1d seqno 45242 age 1 via eno1 neigh fe80::225:90ff:fec1:6252 (installed)
2603:3024:1536:86f0::/64 metric 192 (65535) refmetric 96 id f6:f2:6d:ff:fe:b6:a0:1d seqno 45242 age 1 via eno1 neigh fe80::225:90ff:fec1:6252 (feasible)
fd89:cb20:8854::/48 metric 192 (65535) refmetric 96 id 46:d9:e7:ff:fe:93:82:2d seqno 27834 age 1 via eno1 neigh fe80::225:90ff:fec1:6252 (installed)
babeld: message.c:983: end_message: Assertion `buf->len >= bytes + 2 && buf->buf[buf->len - bytes - 2] == type && buf->buf[buf->len - bytes - 1] == bytes' failed.
Aborted (core dumped)
````

**Comment by dtaht** [comment]
Packet cap...
...
````
172.22.220.0/22 metric 192 (51227) refmetric 96 id 46:d9:e7:ff:fe:93:82:2d seqno 27834 age 1 via eno1 neigh fe80::225:90ff:fec2:2aa3 nexthop 172.22.0.214 (installed)
2601:646:8301:6760::/60 metric 384 (53244) refmetric 288 id a2:21:b7:ff:fe:ac:e4:55 seqno 10074 age 1 via eno1 neigh fe80::225:90ff:fec2:2aa3 (installed)
2603:3024:1536:8600::/64 metric 192 (53117) refmetric 96 id f6:f2:6d:ff:fe:b6:a0:1d seqno 45242 age 1 via eno1 neigh fe80::225:90ff:fec2:2aa3 (installed)
2603:3024:1536:86f0::/60 metric 192 (50652) refmetric 96 id f6:f2:6d:ff:fe:b6:a0:1d seqno 45242 age 1 via eno1 neigh fe80::225:90ff:fec2:2aa3 (installed)
2603:3024:1536:86f0::/64 metric 192 (50513) refmetric 96 id f6:f2:6d:ff:fe:b6:a0:1d seqno 45242 age 1 via eno1 neigh fe80::225:90ff:fec2:2aa3 (feasible)
fd89:cb20:8854::/48 metric 192 (51651) refmetric 96 id 46:d9:e7:ff:fe:93:82:2d seqno 27834 age 1 via eno1 neigh fe80::225:90ff:fec2:2aa3 (installed)
babeld: message.c:983: end_message: Assertion `buf->len >= bytes + 2 && buf->buf[buf->len - bytes - 2] == type && buf->buf[buf->len - bytes - 1] == bytes' failed.

````
[boom.cap.gz](https://github.com/jech/babeld/files/3513175/boom.cap.gz)

**Comment by jech** [comment]
I need a backtrace.

**Comment by dtaht** [comment]
[babeld.gz](https://github.com/jech/babeld/files/3513253/babeld.gz)

**Comment by dtaht** [comment]
it won't let me upload the coredump.

[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `./babeld -d 1 eno1'.
Program terminated with signal SIGABRT, Aborted.
#0  0x00007f5b30c88428 in __GI_raise (sig=sig@entry=6)
    at ../sysdeps/unix/sysv/linux/raise.c:54
54	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  0x00007f5b30c88428 in __GI_raise (sig=sig@entry=6)
    at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007f5b30c8a02a in __GI_abort () at abort.c:89
#2  0x00007f5b30c80bd7 in __assert_fail_base (fmt=<optimized out>, 
    assertion=assertion@entry=0x41590c "buf->len >= bytes + 2 && buf->buf[buf->len - bytes - 2] == type && buf->buf[buf->len - bytes - 1] == bytes", 
    file=file@entry=0x415902 "message.c", line=line@entry=983, 
    function=function@entry=0x416258 <__PRETTY_FUNCTION__.5006> "end_message")
    at assert.c:92
#3  0x00007f5b30c80c82 in __GI___assert_fail (
    assertion=assertion@entry=0x41590c "buf->len >= bytes + 2 && buf->buf[buf->len - bytes - 2] == type && buf->buf[buf->len - bytes - 1] == bytes", 
    file=file@entry=0x415902 "message.c", line=line@entry=983, 
    function=function@entry=0x416258 <__PRETTY_FUNCTION__.5006> "end_message")
    at assert.c:101
#4  0x000000000040b73d in end_message (buf=<optimized out>, 
    type=<optimized out>, bytes=<optimized out>) at message.c:981
#5  0x000000000040d2e5 in send_ihu (neigh=0x20d8180, ifp=<optimized out>)
    at message.c:1742
#6  0x000000000040eded in parse_packet (
    from=from@entry=0x7fff4347f4b8 "\376\200", ifp=0x20d40b0, 
    packet=0x20d41e0 "*\002", packetlen=<optimized out>) at message.c:785
#7  0x0000000000402cd0 in main (argc=<optimized out>, argv=<optimized out>)
    at babeld.c:677

**Comment by jech** [comment]
It appears that you're the first person who tried unicast with RTT estimation.

```
diff --git a/message.c b/message.c
index 270ddc5..0f6ca19 100644
--- a/message.c
+++ b/message.c
@@ -1041,7 +1041,7 @@ buffer_hello(struct buffered *buf, struct interface *ifp,
         accumulate_byte(buf, 4);
         accumulate_int(buf, 0);
     }
-    end_message(&ifp->buf, MESSAGE_HELLO, timestamp ? 12 : 6);
+    end_message(buf, MESSAGE_HELLO, timestamp ? 12 : 6);
 }
 
 void
```

**Comment by jech** [comment]
Fixed in dedc3b899604946b29ad61da5bd93b9451de7242.
